File: web/src/lib/contract-totals.ts
/** Contract subtotal, VAT, and grand total (Saudi VAT default 15%). */

export type ContractTotalsInput = {
  total_rent_amount?: number;
  insurance_amount?: number;
  electricity_amount?: number;
  water_amount?: number;
  services_amount?: number;
  vat_rate?: number;
};

/** Monthly rent × payment count × cycle months (e.g. 12 monthly payments = rent × 1 × 12). */
export function computeTotalRentAmount(
  rentAmount: number,
  paymentCount: number,
  paymentCycle: number,
): number {
  const rent = Number(rentAmount) || 0;
  const count = Number(paymentCount) || 1;
  const cycle = Number(paymentCycle) || 1;
  return Math.round(rent * count * cycle * 100) / 100;
}

export function computeContractTotals(input: ContractTotalsInput) {
  const subtotal =
    (Number(input.total_rent_amount) || 0) +
    (Number(input.insurance_amount) || 0) +
    (Number(input.electricity_amount) || 0) +
    (Number(input.water_amount) || 0) +
    (Number(input.services_amount) || 0);
  const vatRate = Number(input.vat_rate ?? 15) || 0;
  const vatAmount = Math.round(subtotal * (vatRate / 100) * 100) / 100;
  const totalAmount = Math.round((subtotal + vatAmount) * 100) / 100;
  return { subtotal, vatRate, vatAmount, totalAmount };
}

File: api/contract_totals.py
"""Compute contract VAT and grand total from billing line items."""


def compute_contract_totals(
    *,
    total_rent_amount: float,
    insurance_amount: float,
    electricity_amount: float,
    water_amount: float,
    services_amount: float = 0,
    vat_rate: float = 15.0,
) -> tuple[float, float, float]:
    """Return (vat_rate, vat_amount, total_amount)."""
    subtotal = (
        float(total_rent_amount or 0)
        + float(insurance_amount or 0)
        + float(electricity_amount or 0)
        + float(water_amount or 0)
        + float(services_amount or 0)
    )
    rate = float(vat_rate if vat_rate is not None else 15.0)
    vat_amount = round(subtotal * (rate / 100), 2)
    total_amount = round(subtotal + vat_amount, 2)
    return rate, vat_amount, total_amount
